File: pnn/src/nn/engines/trt.rs
use crate::cudnn::{cudnnDataType};
use crate::nn::{DetectionsParser, BoundingBox, RuntimeError, BuildError, Shape, ActivationType};
use crate::cudnn::{DevicePtr};
use crate::nn::Engine;

use std::{
    rc::Rc,
    cell::RefCell,
    collections::HashMap,
    os::raw::{c_void}
};


type Bindings = HashMap<String, DevMemory>;
type DevMemory = Rc<RefCell<DevicePtr>>;

pub struct TRTBuilder {
    // datatype
    data_type: cudnnDataType,
    // Net size, can be usefull for postprocess
    input_size: (usize, usize),
    // Batchsize
    batchsize: usize,
    // Inference operations
    layer_indeces: Vec<usize>,
    // TRTBuilder itself
    builder: *mut c_void
}

fn get_error<T>(msg: &str) -> Result<T, BuildError> {
    Err(BuildError::Runtime(RuntimeError::Other(String::from(msg))))
}

impl TRTBuilder {

    pub fn new(data_type: cudnnDataType, batchsize: usize, input_size: (usize, usize)) -> Result<TRTBuilder, BuildError> {
        let layer_indeces = Vec::new();
        let mut builder = std::ptr::null_mut();
        unsafe {
            builder = pnn_sys::builder_create(data_type.clone() as usize, batchsize);
        }
        if builder == std::ptr::null_mut() {
            return get_error("Couldnt create TRTBuilder")
        }

        Ok(TRTBuilder{batchsize, input_size, data_type, layer_indeces, builder})
    }

    pub fn finilize_layer(&mut self, id: usize) {
        self.layer_indeces.push(id);
    }

    pub fn last_op_id(&self, layer_id: usize) -> usize {
        self.layer_indeces[layer_id]
    }

    pub fn add_input(&mut self, name: &String, shape: Rc<dyn Shape>) -> Result<(), BuildError> {
        let cname = std::ffi::CString::new(name.clone()).unwrap();
        unsafe {
            let id = pnn_sys::builder_add_input(self.builder, 
                cname.as_ptr(), 
                shape.C(),
                shape.H().unwrap(),
                shape.W().unwrap()
            );
            if id < 0 {
                return get_error("Could add Input to TRTBuilder")
            }
            self.layer_indeces.push(id as usize);
        }
        Ok(())
    }

    pub fn add_yolo(&mut self, ind: usize, name: &String) -> Result<(), BuildError> {
        let cname = std::ffi::CString::new(name.clone()).unwrap();
        unsafe {
             pnn_sys::builder_add_yolo(self.builder, 
                ind,
                cname.as_ptr()
            );
        }
        Ok(())
    }

    pub fn add_convolution(&mut self,
        ind: usize,
        filters: usize, 
        input_channels: usize,
        kernel_size: usize,
        padding: usize,
        stride: usize,
        kernels: &Vec<f32>,
        biases: &Vec<f32>
    ) -> Result<usize, BuildError> {
        unsafe {
            let id = pnn_sys::builder_add_convolution(
                self.builder,
                ind,
                filters,
                input_channels,
                kernel_size,
                padding,
                stride,
                kernels.as_ptr(),
                biases.as_ptr()
            );
            if id < 0 {
                return get_error("Could add Convolution to TRTBuilder")
            }
            Ok(id as usize)
        }
    }

    pub fn add_activation(&mut self, ind: usize, act: ActivationType) -> Result<usize, BuildError> {
        let act_name = match act {
            ActivationType::Linear => "linear",
            ActivationType::Mish => "mish",
            ActivationType::Logistic => "logistic"
        };
        let name = std::ffi::CString::new(act_name).unwrap();
        unsafe {
            let id = pnn_sys::builder_add_activation(
                self.builder,
                ind,
                name.as_ptr()
            );
            if id < 0 {
                return get_error("Could add Activation to TRTBuilder")
            }
            Ok(id as usize)
        }
    }

    pub fn add_shortcut(&mut self, indeces: &Vec<usize>) -> Result<usize, BuildError> {
        unsafe {
            let id = pnn_sys::builder_add_shortcut(
                self.builder,
                indeces.len(),
                indeces.as_ptr()
            );
            if id < 0 {
                return get_error("Could add Shortcut to TRTBuilder")
            }
            Ok(id as usize)
        }
    }

    pub fn add_route(&mut self, indeces: &Vec<usize>) -> Result<usize, BuildError> {
        unsafe {
            let id = pnn_sys::builder_add_route(
                self.builder,
                indeces.len(),
                indeces.as_ptr()
            );
            if id < 0 {
                return get_error("Could add Route to TRTBuilder")
            }
            Ok(id as usize)
        }
    }

    pub fn add_pooling(&mut self, ind: usize,
        stride: usize,
        window_size: usize,
        padding: usize,
        is_max: bool
    ) -> Result<usize, BuildError> {       
        unsafe {
            let id = pnn_sys::builder_add_pooling(
                self.builder,
                ind, stride, window_size, padding / 2, is_max as usize
            );
            if id < 0 {
                return get_error("Could add Pooling to TRTBuilder")
            }
            Ok(id as usize)
        }
    }

    pub fn add_upsample(&mut self, ind: usize, stride: usize) -> Result<usize, BuildError> {
        unsafe {
            let id = pnn_sys::builder_add_upsample(self.builder, ind, stride);
            if id < 0 {
                return get_error("Could add Upsample to TRTBuilder")
            }
            Ok(id as usize)
        }
    }

    pub fn build(&mut self, avg_iters: usize, min_iters: usize, engine_path: &String) -> Result<(), BuildError> {
        let name = std::ffi::CString::new(engine_path.clone()).unwrap();
        unsafe {
            let res = pnn_sys::builder_build(
                self.builder,
                avg_iters,
                min_iters,
                name.as_ptr()
            );
            if res == 0 {
                return get_error("Couldnt build engine")
            }
        }
        Ok(())
    }


    pub fn dtype(&self) -> cudnnDataType {
        self.data_type.clone()
    }

    pub fn input_size(&self) -> (usize, usize) {
        self.input_size
    }

}

impl Drop for TRTBuilder {
    fn drop(&mut self) {
        unsafe {
            pnn_sys::builder_destroy(self.builder);
        }
    }
}

pub struct TRTEngine {
    // Batchsize
    batchsize: usize,
    // Inputs
    inputs: Bindings,
    // Outputs
    outputs: Bindings,
    // Parsers for detections
    det_parsers: HashMap<String, Box<dyn DetectionsParser>>,
    // Engine itselft
    engine: *mut c_void,
    // Cuda stream for inference
    stream: crate::cudnn::cudaStream_t,
    // Shapes for bindings, // TODO
}

impl TRTEngine {
    pub fn new(engine_path: &String) -> Result<TRTEngine, RuntimeError> {
        let stream = crate::cudnn::cudaStreamCreate().map_err(|e| {RuntimeError::Cuda(e)})?;
        let mut inputs = HashMap::new();
        let mut outputs = HashMap::new();
        let det_parsers = HashMap::new();
        
        let mut engine = std::ptr::null_mut();
        unsafe {
            let cpath = std::ffi::CString::new(engine_path.clone()).unwrap();
            engine = pnn_sys::engine_create(cpath.as_ptr(), stream);
        }
        if engine == std::ptr::null_mut() {
            return Err(RuntimeError::Other(String::from("Couldnt create TRTEngine")))
        }
        let batchsize = unsafe { pnn_sys::engine_batchsize(engine)};

        let n_bindings = unsafe { pnn_sys::engine_n_bindings(engine)};

        for i in 0..n_bindings {
            let info: pnn_sys::BindingInfo;
            unsafe {
                info = pnn_sys::engine_get_info(engine, i);
            }
            use std::ffi::{CStr, CString};
            let c_name: &CStr = unsafe {CStr::from_ptr(info.name)};
            let name = c_name.to_str().unwrap().to_owned();

            let size = info.batchsize * info.channels * info.height * info.width;
            let dev_ptr = DevicePtr::new(cudnnDataType::FLOAT, size)?;
            unsafe {
                pnn_sys::engine_add_ptr(engine, dev_ptr.ptr());
            }

            let dev_memory = Rc::new(RefCell::new(dev_ptr));
            if info.is_input != 0 {
                inputs.insert(name, dev_memory);
            } else {
                outputs.insert(name, dev_memory);
            }
        }

        Ok(TRTEngine{stream, inputs, outputs, det_parsers, engine, batchsize})
    } 
}

impl Drop for TRTEngine {
    fn drop(&mut self) {
        unsafe {
            pnn_sys::engine_destroy(self.engine);
        }
        crate::cudnn::cudaStreamDestroy(self.stream).unwrap();
    }
}

impl Engine for TRTEngine {

    fn forward(&mut self) -> Result<(), RuntimeError> {
        unsafe {
            pnn_sys::engine_forward(self.engine)
        }
        Ok(())
    }

    fn inputs(&self) -> Vec<String> {
        self.inputs.keys().cloned().collect()
    }

    fn input_binding(&self, name: &String) -> Option<Rc<RefCell<DevicePtr>>> {
        self.inputs.get(name).map(|x| {x.clone()})
    }

    fn output_binding(&self, name: &String) -> Option<Rc<RefCell<DevicePtr>>> {
        self.outputs.get(name).map(|x| {x.clone()})
    }

    fn add_detections_parser(&mut self, binding_name: &String, parser: Box<dyn DetectionsParser>) {
        self.det_parsers.insert(binding_name.clone(), parser);
    }

    fn batchsize(&self) -> usize {
        self.batchsize
    }

    fn detection_parsers(&self) -> &HashMap<String, Box<dyn DetectionsParser>> {
        &self.det_parsers
    }
}

File: pnn-sys/trt/trt.cpp
#include "trt.h"

#include "builder.hpp"
#include "engine.hpp"

void* builder_create(cudnnDataType_t dataType, size_t maxBatchsize) {
    return new TRTBuilder(dataType, maxBatchsize);
}

void builder_destroy(void* builder) {
    delete static_cast<TRTBuilder*>(builder);
}


int builder_add_convolution(void* builder,
    size_t input_id,
    size_t feature_maps,
    size_t input_c,
    size_t kernel_size,
    size_t padding,
    size_t stride,
    float* kernel, 
    float* biases
    ) {
    int64_t params_count = static_cast<int64_t>(feature_maps * input_c * kernel_size * kernel_size);
    Weights wKernel{nvinfer1::DataType::kFLOAT, kernel, params_count};
    Weights wBiases{nvinfer1::DataType::kFLOAT, biases, static_cast<int64_t>(feature_maps)};
    return static_cast<TRTBuilder*>(builder)->addConvolution(input_id, feature_maps, kernel_size, padding, stride, wKernel, wBiases);
}

int builder_add_activation(void* builder, size_t input_id, const char* act_name) {
    return static_cast<TRTBuilder*>(builder)->addActivation(input_id, act_name);
}

int builder_add_shortcut(void* builder, size_t n_layers, size_t *layers) {
    std::vector<size_t> indeces;
    for(size_t i = 0; i < n_layers; i++) indeces.push_back(layers[i]);
    return static_cast<TRTBuilder*>(builder)->addShortcut(indeces);
}

int builder_add_upsample(void* builder, size_t input_id, size_t stride) {
    return static_cast<TRTBuilder*>(builder)->addUpsample(input_id, stride);
}

int builder_add_input(void* builder, const char* name, size_t channels, size_t width, size_t height) {
    return static_cast<TRTBuilder*>(builder)->addInput(name, channels, height, width);
}

void builder_add_yolo(void* builder, size_t input_id, const char* name) {
    static_cast<TRTBuilder*>(builder)->addYolo(input_id, name);
}

int builder_add_route(void* builder, size_t n_layers, size_t *layers) {
    std::vector<size_t> indeces;
    for(size_t i = 0; i < n_layers; i++) indeces.push_back(layers[i]);
    return static_cast<TRTBuilder*>(builder)->addRoute(indeces);
}

int builder_add_pooling(void* builder, size_t input_id, size_t stride, size_t window_size, size_t padding, size_t is_max) {
    return static_cast<TRTBuilder*>(builder)->addPooling(input_id, stride, window_size, padding, is_max);
}

int builder_build(void* builder, size_t avgIters, size_t minIters, const char* engine_path) {
    return static_cast<TRTBuilder*>(builder)->buildEngine(avgIters, minIters, engine_path);
}

void* engine_create(const char* engine_path, cudaStream_t stream) {
    return new TRTEngine(engine_path, stream);
}

void  engine_destroy(void* engine) {
    delete static_cast<TRTEngine*>(engine);
}

size_t engine_batchsize(void* engine) {
    return static_cast<TRTEngine*>(engine)->batchsize();
}

BindingInfo engine_get_info(void* engine, size_t index) {
    return static_cast<TRTEngine*>(engine)->getBindingInfo(index);
}

void engine_add_ptr(void* engine, void* ptr) {
    static_cast<TRTEngine*>(engine)->addBindingPtr(ptr);
}

void engine_forward(void* engine) {
    static_cast<TRTEngine*>(engine)->forward();
}

size_t engine_n_bindings(void* engine) {
    return static_cast<TRTEngine*>(engine)->getNbBindings();
}